doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -14,223 +14,134 @@
-
-Now, span1 and span2 are in the same generation and can both be safely used. After:
-
-.. code-block:: cpp
-
-   arena->freeze();
-   ts::MemSpan span3 = arena->alloc(512);
-   arena->thaw();
-
-span3 can still be used but span1 and span2 have been deallocated and usage is undefined. 
-
-Internal blocks are adjusted for optimization. Each :class:`MemArena::Block` is just a header for the underlying memory it manages. The header and memory are allocated together for locality such that each :class:`MemArena::Block` is immediately followed with the memory it manages. If a :class:`MemArena::Block` is larger than a page (defaulted at 4KB), it is aligned to a power of two. The actual memory that a :class:`MemArena::Block` can allocate out is slightly smaller. This is because a portion of the allocated memory is reserved for the header. Another 16 bytes is reserved to track the allocation headers used by malloc; for page alignment. ie, the default block size is 32768 bytes, but it will only be able to allocate out 32720 bytes. 
+When a :class:`MemArena` instance is constructed it can be constructed with either no allocated
+memory or an initial memory block of a specific size. :func:`MemArena::reserve` can be used to
+specify the size of the next internally allocated block if a specific allocation size is needed.
+This allows allocation to be delayed until memory is actually needed (via a call to
+:func:`MemArena::alloc`).
+
+In normal use memory is allocated from :class:`MemArena` using :func:`MemArena::alloc` to get chunks
+of memory or classes via :func:`MemArena::make` which is effectively :code:`new`.
+:func:`MemArena::make` takes an arbitrary set of arguments which it attempts to pass to a
+constructor for the type :code:`T` after allocating memory for the object. if there isn't enough
+memory a new internal block is allocated, each block being larger than the previous one. The size
+can be controlled using the :func:`MemArena::reserve`. In general it is not as useful to use
+:class:`MemArena` for classes that allocate their own internal memeory (such as :code:`std::string`
+or :code:`std::vector`). One container class that can be easily used is :class:`IntrusiveDList`
+because the links are in the instance and therefore also in the arena. Other than a freeze / thaw
+cycle, there is no mechanism to release memory except for the destruction of the :class:`MemArena`.
+In such use cases either wasted memory must be small enough or temporary enough to not be an issue,
+or there must be a provision for some sort of garbage collection.
+
+After a series of allocations which can be in different allocation blocks with (possibly) unused
+memory in those blocks (either not yet allocated or not longer in use), it can be useful to coalesce
+all of the in use memory in the arena into a single bulk allocation block. This improves memory
+efficiency and memory locality. The methods :func:`MemArena::freeze` and :func:`MemArena::thaw` are
```

Comment:
> After a series of allocations which can be in different allocation blocks with (possibly) unused memory in those blocks (either not yet allocated or not longer in use), it can be useful to coalesce all of the in use memory in the arena into a single bulk allocation block. 

Suggest re-ordering:

```
After a series of allocations it can be useful to coalesce all of the in use memory in the arena into a single bulk allocation block, since memory can be in different allocation blocks with (possibly) unused memory in those blocks (either not yet allocated or not longer in use).
```